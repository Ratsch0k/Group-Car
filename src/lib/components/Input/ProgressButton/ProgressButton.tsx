import React from 'react';
import {
  Button,
  CircularProgress,
  makeStyles,
  createStyles,
  Box,
} from '@material-ui/core';

type Theme = import('@material-ui/core').Theme;
type ButtonProps = import('@material-ui/core').ButtonProps;

export interface ProgressButtonProps extends ButtonProps {
  loading?: boolean;
  progressSize?: number;
}

const ProgressButton: React.FC<ProgressButtonProps> = (props) => {
  const {
    loading,
    children,
    disabled,
    progressSize = 28,
    ...rest
  } = props;

  const useStyle = makeStyles((theme: Theme) =>
    createStyles({
      wrapper: {
        position: 'relative',
      },
      progress: {
        position: 'absolute',
        top: `50%`,
        left: '50%',
        marginLeft: `-${Math.floor(progressSize / 2)}px`,
        marginTop: `-${Math.floor(progressSize / 2)}px`,
      },
    }),
  );

  const classes = useStyle();

  return (
    <Box className={classes.wrapper}>
      <Button {...rest} disabled={disabled || loading}>
        {children}
      </Button>
      {
        loading &&
        <CircularProgress
          className={classes.progress}
          color='secondary'
          size={progressSize}
        />
      }
    </Box>
  );
};

export default ProgressButton;
