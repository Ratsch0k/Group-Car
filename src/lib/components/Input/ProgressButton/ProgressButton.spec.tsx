import React from 'react';
import ProgressButton from './ProgressButton';
import {screen, render} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

it('renders without crashing', () => {
  render(<ProgressButton />);
});

it('renders with button disabled and circular progress if loading', () => {
  const {container} = render(<ProgressButton loading={true} />);
  expect(screen.getByRole('button')).toHaveAttribute('disabled');
  expect(screen.getByRole('progressbar')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

it('renders with button not disabled ' +
    'and with no pr ogress if not loading', () => {
  const {container} = render(<ProgressButton loading={false} />);
  expect(screen.getByRole('button')).not.toHaveAttribute('disabled');
  expect(screen.queryByRole('progressbar')).toBeNull();
  expect(container).toMatchSnapshot();
});
