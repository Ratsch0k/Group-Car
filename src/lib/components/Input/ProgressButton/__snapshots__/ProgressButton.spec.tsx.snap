// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with button disabled and circular progress if loading 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-50 makeStyles-wrapper-47"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
    </button>
    <div
      class="MuiCircularProgress-root makeStyles-progress-48 MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 28px; height: 28px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`renders with button not disabled and with no pr ogress if not loading 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-98 makeStyles-wrapper-95"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
